upgrades.go

```diff
@@ -0,0 +1,186 @@
+		if err != nil {
+			log.Error(err)
+		}
+		if isNew {
+			log.Info("Disabled OS auto-upgrades.")
+		}
+		return
+	}
+
+	err = verifyInstall()
+	if err != nil {
+		err = doInstall(ctx)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+	}
+
+	securityOnly := cfg.Type == "security"
+	confContents, err := generateOrigins(securityOnly)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	isNew1, err := writeFileIfNew(autoUpgradesPath, []byte(autoUpgradesContentsEnabled))
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	isNew2, err := writeFileIfNew(unattendedUpgradesPath, []byte(confContents))
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if isNew1 || isNew2 {
+		if securityOnly {
+			log.Info("Enabled OS auto-upgrades (security only.)")
+		} else {
+			log.Info("Enabled OS auto-upgrades (full.)")
+		}
+	}
+
+	err = enableTimer()
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+func checkSupportedDistro() error {
+	data, err := os.ReadFile("/etc/os-release")
+	if err != nil {
+		return err
+	}
+
+	if strings.Contains(string(data), "VERSION_CODENAME=bookworm") || strings.Contains(string(data), "VERSION_CODENAME=bullseye") {
+		return nil
+	}
+
+	return errw.New("cannot enable automatic upgrades for unknown distro, only support for Debian bullseye and bookworm is available")
+}
+
+// make sure the needed package is installed.
+func verifyInstall() error {
+	cmd := exec.Command("unattended-upgrade", "-h")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return errw.Wrapf(err, "executing 'unattended-upgrade -h' %s", output)
+	}
+	return nil
+}
+
+func enableTimer() error {
+	// enable here
+	cmd := exec.Command("systemctl", "enable", "apt-daily-upgrade.timer")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return errw.Wrapf(err, "executing 'systemctl enable apt-daily-upgrade.timer' %s", output)
+	}
+	return nil
+}
+
+func doInstall(ctx context.Context) error {
+	// On low bandwidth connections, apt updates/installs can take a while, so start something to handle healthchecks
+	sleepCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		HealthySleep(sleepCtx, time.Hour)
+	}()
+
+	cmd := exec.CommandContext(ctx, "apt", "update")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return errw.Wrapf(err, "executing 'apt update' %s", output)
+	}
+
+	cmd = exec.CommandContext(ctx, "apt", "install", "-y", "unattended-upgrades")
+	output, err = cmd.CombinedOutput()
+	if err != nil {
+		return errw.Wrapf(err, "executing 'apt install -y unattended-upgrades' %s", output)
+	}
+
+	return nil
+}
+
+// generates the "Origins-Pattern" section of 50unattended-upgrades file.
+func generateOrigins(securityOnly bool) (string, error) {
+	cmd := exec.Command("apt-cache", "policy")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errw.Wrapf(err, "executing 'apt-cache policy' %s", output)
+	}
+
+	releaseRegex := regexp.MustCompile(`release.*o=([^,]+).*n=([^,]+).*`)
```

Comment:
optionally add a unit test for the string-mashing parts of this function; confident in correctness today but imagine if *I* have to modify this is a month. I will definitely break it

Reply:
Unit tests? UNIT TESTS! What is this, proper development?

I'm kidding, but yes, we need a variety of tests for agent as  a whole. Ale and I have discussed we probably need a "human testing" document, and units tests for stuff like this too.